File: LeetCode/Go/515-breadth-first-search-在每个树行中找最大值.go
/*
 * @lc app=leetcode.cn id=515 lang=golang
 *
 * [515] 在每个树行中找最大值
 */
package leetcode

import "math"

// @lc code=start
func largestValues(root *TreeNode) []int {
	res := []int{}
	q := []*TreeNode{}
	if root != nil {
		q = append(q, root)
	}
	for len(q) != 0 {
		res = append(res, math.MinInt64)
		size := len(q)
		for i := 0; i < size; i++ {
			front := q[0]
			q = q[1:]
			if front.Val > res[len(res)-1] {
				res[len(res)-1] = front.Val
			}
			if front.Left != nil {
				q = append(q, front.Left)
			}
			if front.Right != nil {
				q = append(q, front.Right)
			}
		}
	}
	return res
}

// @lc code=end

File: LeetCode/Java/515-breadth-first-search-在每个树行中找最大值.java
import java.util.*;
/*
 * @lc app=leetcode.cn id=515 lang=java
 *
 * [515] 在每个树行中找最大值
 */

// @lc code=start
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        var res = new ArrayList<Integer>();

        if (root == null)
            return res;

        Queue<TreeNode> q = new LinkedList();
        q.add(root);

        while (!q.isEmpty()) {
            int size = q.size();
            int max = Integer.MIN_VALUE;
            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                max = Math.max(max, node.val);
                if (node.left != null)
                    q.add(node.left);
                if (node.right != null)
                    q.add(node.right);
            }
            res.add(max);
        }

        return res;
    }
}
// @lc code=end
